src/templater.rs

```diff
@@ -481,18 +508,23 @@ impl FormattedCommitOrChangeId<'_> {
 
     pub fn short(&self) -> String {
         let mut hex = self.hex();
-        hex.truncate(12);
+        if let Some(len) = self.trim_display_to {
+            hex.truncate(len);
+        }
         hex
     }
 
     pub fn shortest_prefix_and_brackets(&self) -> String {
-        highlight_shortest_prefix_brackets(self, 12)
+        highlight_shortest_prefix_brackets(self, self.trim_display_to)
```

Comment:
Can you add a separate variable for the `shortest` `trim_display_to`?

This `trim_display_to` behaves differently from the `short` `trim_display_to`, and that's my main concern about this patch.

Other than that, I think this PR is good to go, and I really like this feature, thanks.

Reply:
What should the logic be if both variables are set?

Reply:
> What should the logic be if both variables are set?

I think the current behavior (use short if the PrefixFormat is None, otherwise use shortest) is okay. 

And that's why I considered splitting structs (or types) would be better. No need to do that in this PR, though.

Reply:
Ah, it's probably easier to make `id.short()` return String as before. Then, `FormattedCommitOrChangeId` doesn't have to deal with the "short" trim_display_to.

Reply:
Sure, I can do that. I'll probably rename FormattedCommitChangeId to something including the word prefix.

I'll confess that it's not clear to me why this matters to you, to me it feels neither worse not better. Perhaps it'll become clearer with time.

Reply:
Maybe I'm not a good writer.

Some artificial example: `-T 'commit_id.short().shortest_styled_prefix().shortest_prefix_and_brackets()'`
What's the expected output for this template? I don't care as a user, but I expect there is some consistent behavior. That's the motivation to make `short()` return String, (and `.shortest()` return something else.)

And I think separate short/shortest `trim_display_to`s would make clear that they behave differently, and thus such method chaining doesn't have a consistent behavior.

Reply:
At this point, I think we settled on a solution that everyone is happy with. I won't get to implementing it today, but it's no trouble and I'll do it soon.

> That's the motivation to make short() return String

To make sure we're on the same page, this should result in `-T 'commit_id.short().shortest_styled_prefix().shortest_prefix_and_brackets()'` being an error.

> What's the expected output for this template?

I might be biased by what my code did, but it would make sense to me for the last of these functions to win. At least, I don't see this behavior as worse than the alternatives.

Note that I think that your suggestion of making `short` a string makes just as much sense. I trust your intuition and I won't be surprised if one day the difference will turn out to matter in some way. Even if that doesn't happen, it's not a problem.

Reply:
> > That's the motivation to make short() return String
> 
> To make sure we're on the same page, this should result in `-T 'commit_id.short().shortest_styled_prefix().shortest_prefix_and_brackets()'` being an error.

Yes.

> I might be biased by what my code did, but it would make sense to me for the last of these functions to win. At least, I don't see this behavior as worse than the alternatives.

Yeah, that's another way around. If we took that direction, internal API could be `.set_format(prefix_format, trim_display_to)` to ensure that the last `.short()` win with
`'commit_id.shortest_styled_prefix().short()'`.

> Note that I think that your suggestion of making `short` a string makes just as much sense.

+1

Reply:
I just realized that making `short()` a String does not remove the need for separate `CommitOrChangeId` and `HighlightedPrefixCommitOrChangeId` types inside `Property`, since it doesn't make much sense for `commit_id` to have the `HighlightedPrefixCommitOrChangeId` type before any function is called on it.

So, I think I'll make something closer to your suggestion from https://github.com/martinvonz/jj/pull/1100#discussion_r1089936894.

Reply:
Think `Property::<Type>(Box<... Output = <Type>>)` and the related match arms are generated codes. We have to be a code generator atm, but I think it can be ported to some kind of macro rules.

Reply:
OK, here is the converted version. PTAL.